[WIP] storage: proposal quota pool

petermattis opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13869)
<!-- Reviewable:end -->

petermattis committed:
[WIP] storage: proposal quota pool

The leader maintains a pool of "proposal quota". Before proposing a Raft
command, we acquire 1 unit of proposal quota. When all of the active
followers have committed an entry, that unit of proposal quota is
returned to the pool. The proposal quota pool size is hard coded to 1000
which allows fairly deep pipelining of Raft commands.

We only consider "active" followers when determining if a unit of quota
should be returned to the pool. An active follower is one we've received
any type of message from in the past 2 seconds.

See #8659

petermattis commented:
@spencerkimball Per our discussion yesterday. This isn't fully fleshed out right now, but gives us something concrete to argue over.

petermattis commented:
This seems to work in simple testing. A node crash will cause a 2 second blip of unavailability if the range is very active. In normal operation, we'll limit proposals to the speed of the slowest replica, but with a buffer to avoid imposing the latency of the slower replicas on the commit latency.

Cc @bdarnell

spencerkimball commented on pkg/storage/quota_pool.go:
```diff
@@ -0,0 +1,81 @@
+)
+
+const (
+	leaderProposalQuota   = 1000
```
Do we need something more thoughtful here?

petermattis replied:
Probably. We'd like something that avoided Raft log truncation. This achieves that for the workload I'm using which involves small requests. More directly estimating the affect on Raft log size would be nice, though I'm not sure how to achieve that.

spencerkimball commented on pkg/storage/quota_pool.go:
```diff
@@ -0,0 +1,81 @@
+type quotaPool struct {
+	c chan int64
+
+	mu    sync.Mutex
```
Any reason not to use `atomic` primitives?

petermattis replied:
This code was lifted from `grpc`. I haven't thought through the difficulties with using atomics here.

spencerkimball replied:
yep, could be tricky. You'll have to scrutinize the code carefully.

spencerkimball commented on pkg/storage/quota_pool.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+func newQuotaPool(q int64) *quotaPool {
+	qb := &quotaPool{
```
Why is this `qb`, not `qp`?

petermattis replied:
Cause I stole it from the `grpc` code base which uses the name `qb`. I'll change it eventually.

spencerkimball commented on pkg/storage/replica.go:
```diff
@@ -2749,7 +2793,17 @@ func (r *Replica) handleRaftReadyRaftMuLocked(
+			r.mu.proposalQuotaBase = r.mu.lastIndex
+			r.proposalQuota.reset(leaderProposalQuota)
+		} else {
+			r.proposalQuota.reset(followerProposalQuota)
```
Why do followers require a proposal quota?

petermattis replied:
So they can propose requests which get forwarded to the leader. For example, followers propose lease requests which is how they determine who holds the lease. We could probably change this so that we don't acquire proposal quota when a follower proposes a Raft command.

spencerkimball replied:
I'd just as soon eliminate the follower's use of a quota pool and the associated constant and just set the pool to nil on a leader -> follower transition, and allow followers to always propose and forward to the leader.

spencerkimball commented on pkg/storage/replica.go:
```diff
@@ -420,11 +421,15 @@ type Replica struct {
 		// Computed checksum at a snapshot UUID.
 		checksums map[uuid.UUID]replicaChecksum
 
+		proposalQuotaBase uint64
```
s/proposalQuotaBase/proposalQuotaBaseIndex/

petermattis replied:
Done.

spencerkimball commented on pkg/storage/store.go:
```diff
@@ -3096,6 +3096,7 @@ func (s *Store) processRaftRequest(
 		if req.Message.Type == raftpb.MsgApp {
 			r.setEstimatedCommitIndexLocked(req.Message.Commit)
 		}
+		r.setLastActivityLocked(req.FromReplica.ReplicaID)
```
Is it possible for a hopelessly lagged replica to come back online and require a snapshot, but still report its log index and cause the quota pool to be immediately drained to a negative value?

petermattis replied:
No, that's not possible. The purpose of `Replica.mu.proposalQuotaBaseIndex` is that the index used for quota calculations ratchets up. If a replica comes back on line and has a commit index before the quota base index, it will be ignored for quota purposes until it catches back up.

spencerkimball replied:
Ah, right.

petermattis replied:
Actually, this was broken before, but should be fixed now.

spencerkimball commented on pkg/storage/replica.go:
```diff
@@ -800,6 +806,40 @@ func (r *Replica) setEstimatedCommitIndexLocked(commit uint64) {
+	for _, rep := range r.mu.state.Desc.Replicas {
+		// Only consider followers that we've received a message from in the last 2
+		// seconds.
+		const activeTime = 2 * time.Second
```
Probably need to put this in `StoreConfig`.

petermattis replied:
Yep. I'll add a TODO for now.

spencerkimball reviewed (commented):
Overall this is pretty simple, which is positive. As mentioned in the review comments, I think the biggest concern would be a replica coming back online and wreaking havoc.

My earlier thought to avoid the 2s timeout seems less useful now that I'm looking at this. It would just push the problem I was looking to solve elsewhere – we'd be relying on node liveness instead of this 2s timeout.

Need unittests too, obviously.

petermattis reviewed (commented):
Yes, unit tests are definitely needed before merging.

spencerkimball commented:
Aside from unittests, LGTM

petermattis commented:
I reworked the handling of `quotaPool` so that it uses `Replica.mu` instead of its own internal lock and I changed the code so that there is only a non-nil `quotaPool` on the leader.

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 2802 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-Ke8pGIn0GfcQX_OaDIb-r1-2802:-KeAKa0DbixflGKHmo0q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e031a0bed284bd6fffc954ea66f65888ef4f737/pkg/storage/replica.go#L2802)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

I'd just as soon eliminate the follower's use of a quota pool and the associated constant and just set the pool to nil on a leader -> follower transition, and allow followers to always propose and forward to the leader.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/storage/quota_pool.go, line 18 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-Ke8pGIn0GfcQX_OaDIc-r1-18:-KeAKg6wt4LlFeJmmqyO:b97644t) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e031a0bed284bd6fffc954ea66f65888ef4f737/pkg/storage/quota_pool.go#L18)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

yep, could be tricky. You'll have to scrutinize the code carefully.
</blockquote></details>

Reworked to remove the mutex.

---

*[pkg/storage/quota_pool.go, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-Ke8pGIn0GfcQX_OaDIc-r1-23:-KeAKev3GWogEOqwFcKO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e031a0bed284bd6fffc954ea66f65888ef4f737/pkg/storage/quota_pool.go#L23)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Cause I stole it from the `grpc` code base which uses the name `qb`. I'll change it eventually.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/storage/quota_pool.go, line 18 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-Ke8pGIn0GfcQX_OaDIc-r1-18:-KeAQsF9dCHgYvGhMnPI:b-zhtrqw) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e031a0bed284bd6fffc954ea66f65888ef4f737/pkg/storage/quota_pool.go#L18)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Reworked to remove the mutex.
</blockquote></details>

Doh! Removing the mutex clearly wasn't safe. Reintroduced for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 1 of 3 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[pkg/storage/quota_pool.go, line 75 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVaV8LR9msq-s7zujX:-KeVaV8LR9msq-s7zujY:b-6qafsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/quota_pool.go#L75)):*
> ```Go
> 		qp.c <- q
> 	} else {
> 		qp.quota = q
> ```

Is `q` always zero here (in which case this assignment is redundant) or are negative values allowed?

---

*[pkg/storage/quota_pool.go, line 101 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVajdXZFizFynWFsIe:-KeVajdXZFizFynWFsIf:bx4fnyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/quota_pool.go#L101)):*
> ```Go
> 
> // acquires acquires a single unit of quota from the pool. Safe for concurrent
> // use.
> ```

Document what is required of the caller: if acquire() returns without error, the caller must call add(1) after doing its work. 

---

*[pkg/storage/replica.go, line 825 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-Ke8pGIn0GfcQX_OaDIb-r1-825:-KeVnS5nnLegZR3Cv8vF:b-uhkpl5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e031a0bed284bd6fffc954ea66f65888ef4f737/pkg/storage/replica.go#L825)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Yep. I'll add a TODO for now.
</blockquote></details>

We have a lot of different failure detectors, with different thresholds and even different mechanisms. Can this piggyback on another (probably NodeLiveness) instead of introducing a new way for a node to be "down"? 

A failing node already triggers a blip of up to 9s (?) for ranges it leads, so it doesn't seem bad to me to lengthen the impact of a failing node on ranges where it is a follower which are also maxing out their proposal quota. 

---

*[pkg/storage/replica.go, line 812 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVbosLgJ5DyjJEZ33u:-KeVbosLgJ5DyjJEZ33v:bgqt5dm) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/replica.go#L812)):*
> ```Go
> 		r.mu.lastActivity = make(map[roachpb.ReplicaID]time.Time)
> 	}
> 	r.mu.lastActivity[replicaID] = timeutil.Now()
> ```

We should also clear from this map at some point - at least when removing a replica, and maybe also empty the whole map when leadership changes hands? 

---

*[pkg/storage/replica.go, line 825 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVcIDP2zhKc8-dvXKL:-KeVcIDP2zhKc8-dvXKM:b-6u4hde) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/replica.go#L825)):*
> ```Go
> }
> 
> func (r *Replica) refreshProposalQuotaLocked(newLeaderID roachpb.ReplicaID) {
> ```

s/refresh/update/?

---

*[pkg/storage/replica.go, line 840 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVeBjjzpZbQf3wlNdq:-KeVeBjjzpZbQf3wlNdr:b-nbovu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/replica.go#L840)):*
> ```Go
> 		return
> 	}
> 	// We're the leader.
> ```

Add "still" to this comment to distinguish it from the early return above when we're becoming the leader. 

---

*[pkg/storage/replica.go, line 842 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVknK6H4GNctpDptGV:-KeVknK79PMg_B635TRX:b85o6r0) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/replica.go#L842)):*
> ```Go
> 	// We're the leader.
> 
> 	// TODO(peter): Can we avoid retrieving the Raft status on every invocation?
> ```

I assume your concern is for the allocations this performs? We could probably add a method to query this without additional allocations. We could also maintain our own map by tracking MsgAppResps as they pass through the transport, but that probably duplicates too much raft logic. 

---

*[pkg/storage/replica.go, line 844 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVkO--SVekOjDZnxgk:-KeVkO--SVekOjDZnxgl:bdxazkj) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/replica.go#L844)):*
> ```Go
> 	// TODO(peter): Can we avoid retrieving the Raft status on every invocation?
> 	status := r.raftStatusRLocked()
> 	// Find the minimum index that active followers have committed.
> ```

s/committed/acknowledged/

---

*[pkg/storage/replica.go, line 857 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVkDJPx2wQMIIy36YB:-KeVkDJPx2wQMIIy36YC:b84t8ls) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/replica.go#L857)):*
> ```Go
> 		}
> 		if progress, ok := status.Progress[uint64(rep.ReplicaID)]; ok {
> 			// Only consider follower's who are in advance of the quota base
> ```

`s/'//`

---

*[pkg/storage/replica.go, line 870 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVmF9dVuCVhr3ElKJg:-KeVmF9dVuCVhr3ElKJh:b78nfdk) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/replica.go#L870)):*
> ```Go
> 
> 	if r.mu.proposalQuotaBaseIndex < minIndex {
> 		delta := int64(minIndex - r.mu.proposalQuotaBaseIndex)
> ```

Add a comment pointing out this subtlety: Raft may propose commands itself (specifically the empty commands when leadership changes), and these commands don't go through the code paths where we acquire quota from the pool. We avoid releasing quota here that we never acquired by resetting the quota pool whenever leadership changes hands. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I still need to add tests.

---

Review status: 1 of 3 files reviewed at latest revision, 16 unresolved discussions.

---

*[pkg/storage/quota_pool.go, line 75 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVaV8LR9msq-s7zujX:-KeWJLZmJe-iMoQZchu0:b-qctosc) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/quota_pool.go#L75)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Is `q` always zero here (in which case this assignment is redundant) or are negative values allowed?
</blockquote></details>

We always pass in a positive value to `newQuotaPool`. I can simplify this. 

---

*[pkg/storage/quota_pool.go, line 101 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVajdXZFizFynWFsIe:-KeWLfc2aPClIGicHYDA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/quota_pool.go#L101)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Document what is required of the caller: if acquire() returns without error, the caller must call add(1) after doing its work. 
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 825 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-Ke8pGIn0GfcQX_OaDIb-r1-825:-KeWLw3WtTNzd_rMznq0:b-3ugpfy) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e031a0bed284bd6fffc954ea66f65888ef4f737/pkg/storage/replica.go#L825)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We have a lot of different failure detectors, with different thresholds and even different mechanisms. Can this piggyback on another (probably NodeLiveness) instead of introducing a new way for a node to be "down"? 

A failing node already triggers a blip of up to 9s (?) for ranges it leads, so it doesn't seem bad to me to lengthen the impact of a failing node on ranges where it is a follower which are also maxing out their proposal quota. 
</blockquote></details>

I initially piggybacked this on node liveness, but then got worried that the node liveness ranges would have to be exempted in order to avoid deadlock. Seemed simpler to have a different mechanism here.

---

*[pkg/storage/replica.go, line 812 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVbosLgJ5DyjJEZ33u:-KeWMfQ6ct77XFmaBhq6:bvnvjix) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/replica.go#L812)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We should also clear from this map at some point - at least when removing a replica, and maybe also empty the whole map when leadership changes hands? 
</blockquote></details>

Done. We now only initialize this map on the leader and clear it when a node becomes a follower.

---

*[pkg/storage/replica.go, line 825 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVcIDP2zhKc8-dvXKL:-KeWMekTsColVx6sy7i7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/replica.go#L825)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/refresh/update/?
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 840 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVeBjjzpZbQf3wlNdq:-KeWMqVnbTvHdlmxp6XE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/replica.go#L840)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Add "still" to this comment to distinguish it from the early return above when we're becoming the leader. 
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 842 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVknK6H4GNctpDptGV:-KeWMrJsiOmyO9dA_4pl:brwq17q) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/replica.go#L842)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I assume your concern is for the allocations this performs? We could probably add a method to query this without additional allocations. We could also maintain our own map by tracking MsgAppResps as they pass through the transport, but that probably duplicates too much raft logic. 
</blockquote></details>

Yeah, the allocation is the concern. It is a small concern, though. Trying to track this ourselves via MsgAppResp seems like overkill. I've enhanced the TODO comment.

---

*[pkg/storage/replica.go, line 844 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVkO--SVekOjDZnxgk:-KeWNDui9wTMcH-lH3AK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/replica.go#L844)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/committed/acknowledged/
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 857 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVkDJPx2wQMIIy36YB:-KeWNHsS09ZJmtTJklRx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/replica.go#L857)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`s/'//`
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 870 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVmF9dVuCVhr3ElKJg:-KeWNR8wvQR26__QBCe5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/replica.go#L870)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Add a comment pointing out this subtlety: Raft may propose commands itself (specifically the empty commands when leadership changes), and these commands don't go through the code paths where we acquire quota from the pool. We avoid releasing quota here that we never acquired by resetting the quota pool whenever leadership changes hands. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-:-KeWNWf0muMqvu106xPA:b-nna9ol)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 3 files reviewed at latest revision, 16 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 825 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-Ke8pGIn0GfcQX_OaDIb-r1-825:-KeWORBLbCrqqgn3OuR6:b-ou6bw2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e031a0bed284bd6fffc954ea66f65888ef4f737/pkg/storage/replica.go#L825)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I initially piggybacked this on node liveness, but then got worried that the node liveness ranges would have to be exempted in order to avoid deadlock. Seemed simpler to have a different mechanism here.
</blockquote></details>

Rather than introducing a new configurable here, perhaps this time should be the same as the Raft election timeout. The only way for a leader to have not heard for a Replica for longer than the Raft election timeout is for the Range to be quiescent in which case there are no concerns about proposal quota. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 3 files reviewed at latest revision, 16 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 825 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-Ke8pGIn0GfcQX_OaDIb-r1-825:-KeWOvMYbRmqrKHER8Ms:b-56qcyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e031a0bed284bd6fffc954ea66f65888ef4f737/pkg/storage/replica.go#L825)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Rather than introducing a new configurable here, perhaps this time should be the same as the Raft election timeout. The only way for a leader to have not heard for a Replica for longer than the Raft election timeout is for the Range to be quiescent in which case there are no concerns about proposal quota. 
</blockquote></details>

Err, `The only way` is not correct. `The primary way`? A leader might not have heard from a replica if that replica is on a down/partitioned node.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 825 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-Ke8pGIn0GfcQX_OaDIb-r1-825:-KeZRvh8-S7aonTEd3o8:bsiz0op) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e031a0bed284bd6fffc954ea66f65888ef4f737/pkg/storage/replica.go#L825)):*
> but then got worried that the node liveness ranges would have to be exempted in order to avoid deadlock

OK, what about RPC heartbeats then? I'd really like to avoid introducing a new notion of node activity, especially one with subtle interactions with quiescence and leadership changes. 

---

*[pkg/storage/replica.go, line 812 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-KeVbosLgJ5DyjJEZ33u:-KeZQvXFz6woGpfC_sbi:bbcmc8k) ([raw file](https://github.com/cockroachdb/cockroach/blob/813d7bcf8aa44bd8196d47c72ce1c25f94fe5e8d/pkg/storage/replica.go#L812)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Done. We now only initialize this map on the leader and clear it when a node becomes a follower.
</blockquote></details>

So a brand-new leader will see all the followers as inactive. That seems less than ideal, although since we're also resetting the quota to its maximum limit when becoming leader I don't see a specific problem with it. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869)*
<!-- Sent from Reviewable.io -->

petermattis committed:
storage: replace lastActivity with RPC heartbeats

petermattis commented:
Review status: 1 of 2 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/storage/replica.go, line 825 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869#-Ke8pGIn0GfcQX_OaDIb-r1-825:-KeZi7DevlBphFXAvWPn:bph424k) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e031a0bed284bd6fffc954ea66f65888ef4f737/pkg/storage/replica.go#L825)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

> but then got worried that the node liveness ranges would have to be exempted in order to avoid deadlock

OK, what about RPC heartbeats then? I'd really like to avoid introducing a new notion of node activity, especially one with subtle interactions with quiescence and leadership changes. 
</blockquote></details>

My instinct is that the new mechanism would be less fragile than reusing an existing one like RPC heartbeats. Not sure why I feel that way. I'll think about this more. That said, I added another commit which replaces `lastActivity` with RPC connection health. The various lookups involved are mildly concerning for performance. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13869)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Cc @irfansharif
